### PR TITLE
Configuration file from environment variable

### PR DESCRIPTION
**Pull request by p1c2u**
Adds support for setting configuration file from `ALEMBIC_CONFIG` environment variable

### CODE REVIEW

**Comment by zzzeek** [comment]
hi there - 

This is not stated in the PR guidelines for alembic but I very much prefer that an issue is opened first for a new feature so that the rationale may be discussed.   That way you don't have to go through all the trouble of writing a PR with tests (looks great btw except it would need documentation) in case we decide this is not how we want to actually solve this problem.

in the test I see some pushing of values into config but what's the full rationale for this feature ?    

I assume what you are going for is to impact the command-line environment first and foremost:

    alembic -c <config_file>

becomes:

  ALEMBIC_CONFIG=<config_file> alembic

is that it?     Is there a reason this is not sufficient:

   alembic -c ${ALEMBIC_CONFIG}

**Comment by p1c2u** [comment]
Hi @zzzeek . Thank you for your comment. I created issue for the PR as you suggested.

**Comment by sqla-tester** [comment]
New Gerrit review created for change 0ea7f8fea69f2be5f9ce5be2343149bf62f0dfcb: https://gerrit.sqlalchemy.org/#/c/sqlalchemy/alembic/+/1502

**Comment by zzzeek** [comment]
IIUC this will not work if ALEMBIC_CONFIG is set and then someone passes -c as well, -c should supersede the environment variable

File: alembic/config.py [hunk 1]
```diff
@@ -103,8 +109,10 @@ def __init__(
         """Construct a new :class:`.Config`
 
         """
-        self.config_file_name = file_
```
Comment (zzzeek) [inline alembic/config.py]: this is all very complicated, and won't work if someone specifies -c, why can't isnt ALEMBIC_CONFIG simply consumed inside of CommandLine.main() when we pass the name of the config file to the Config (one line change basically?)  is there some reason you need programmatic access to os.environ within the Config object itself  as well as the ability to make use of an alternate dictionary to os.environ ?

**Comment by zzzeek** [comment]
does this patch work for you ?

```diff
diff --git a/alembic/config.py b/alembic/config.py
index b0700d5..22d8152 100644
--- a/alembic/config.py
+++ b/alembic/config.py
@@ -468,12 +468,14 @@ class CommandLine(object):
                     subparser.add_argument(arg, help=positional_help.get(arg))
 
         parser = ArgumentParser(prog=prog)
+
         parser.add_argument(
             "-c",
             "--config",
             type=str,
-            default="alembic.ini",
-            help="Alternate config file",
+            default=os.environ.get("ALEMBIC_CONFIG", "alembic.ini"),
+            help="Alternate config file; defaults to value of "
+            "ALEMBIC_CONFIG environment variable, or \"alembic.ini\"",
         )
         parser.add_argument(
             "-n",

```

**Comment by zzzeek** [comment]
OK I have this proposed in  https://gerrit.sqlalchemy.org/1514

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

going with  https://gerrit.sqlalchemy.org/1514 for now

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/1502 has been **abandoned**.  That means that at least for the moment I need to close this pull request. Sorry it didn't work out :(

**Pull request closed without merging**